// avalanche/typings/tests/apis/ipcs/api.test.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jest_mock_axios_1 = __importDefault(require("jest-mock-axios"));
const src_1 = require("src");
describe('IPCS', () => {
    const ip = '127.0.0.1';
    const port = 9650;
    const protocol = 'https';
    const avalanche = new src_1.Avalanche(ip, port, protocol, 12345);
    let ipcs;
    beforeAll(() => {
        ipcs = avalanche.IPCS();
    });
    afterEach(() => {
        jest_mock_axios_1.default.reset();
    });
    test('publishBlockchain', () => __awaiter(void 0, void 0, void 0, function* () {
        const blockchainID = "11111111111111111111111111111111LpoYY";
        const result = ipcs.publishBlockchain(blockchainID);
        const payload = {
            result: {
                consensusURL: "/tmp/12345-11111111111111111111111111111111LpoYY-consensus",
                decisionsURL: "/tmp/12345-11111111111111111111111111111111LpoYY-decisions"
            },
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response.consensusURL).toBe("/tmp/12345-11111111111111111111111111111111LpoYY-consensus");
        expect(response.decisionsURL).toBe("/tmp/12345-11111111111111111111111111111111LpoYY-decisions");
    }));
    test('unpublishBlockchain', () => __awaiter(void 0, void 0, void 0, function* () {
        const blockchainID = "11111111111111111111111111111111LpoYY";
        const result = ipcs.unpublishBlockchain(blockchainID);
        const payload = {
            result: {
                success: true
            },
        };
        const responseObj = {
            data: payload
        };
        jest_mock_axios_1.default.mockResponse(responseObj);
        const response = yield result;
        expect(jest_mock_axios_1.default.request).toHaveBeenCalledTimes(1);
        expect(response).toBe(true);
    }));
});
