// avalanche/dist/apis/info/api.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InfoAPI = void 0;
const jrpcapi_1 = require("../../common/jrpcapi");
const bn_js_1 = __importDefault(require("bn.js"));
/**
 * Class for interacting with a node's InfoAPI.
 *
 * @category RPCAPIs
 *
 * @remarks This extends the [[JRPCAPI]] class. This class should not be directly called. Instead, use the [[Avalanche.addAPI]] function to register this interface with Avalanche.
 */
class InfoAPI extends jrpcapi_1.JRPCAPI {
    constructor(core, baseurl = '/ext/info') {
        super(core, baseurl);
        /**
         * Fetches the blockchainID from the node for a given alias.
         *
         * @param alias The blockchain alias to get the blockchainID
         *
         * @returns Returns a Promise<string> containing the base 58 string representation of the blockchainID.
         */
        this.getBlockchainID = (alias) => __awaiter(this, void 0, void 0, function* () {
            const params = {
                alias,
            };
            return this.callMethod('info.getBlockchainID', params)
                .then((response) => response.data.result.blockchainID);
        });
        /**
         * Fetches the networkID from the node.
         *
         * @returns Returns a Promise<number> of the networkID.
         */
        this.getNetworkID = () => __awaiter(this, void 0, void 0, function* () {
            const params = {};
            return this.callMethod('info.getNetworkID', params)
                .then((response) => response.data.result.networkID);
        });
        /**
         * Fetches the network name this node is running on
         *
         * @returns Returns a Promise<string> containing the network name.
         */
        this.getNetworkName = () => __awaiter(this, void 0, void 0, function* () {
            return this.callMethod('info.getNetworkName')
                .then((response) => response.data.result.networkName);
        });
        /**
         * Fetches the nodeID from the node.
         *
         * @returns Returns a Promise<string> of the nodeID.
         */
        this.getNodeID = () => __awaiter(this, void 0, void 0, function* () {
            const params = {};
            return this.callMethod('info.getNodeID', params)
                .then((response) => response.data.result.nodeID);
        });
        /**
         * Fetches the version of Gecko this node is running
         *
         * @returns Returns a Promise<string> containing the version of Gecko.
         */
        this.getNodeVersion = () => __awaiter(this, void 0, void 0, function* () {
            return this.callMethod('info.getNodeVersion')
                .then((response) => response.data.result.version);
        });
        /**
         * Fetches the transaction fee from the node.
         *
         * @returns Returns a Promise<object> of the transaction fee in nAVAX.
         */
        this.getTxFee = () => __awaiter(this, void 0, void 0, function* () {
            return this.callMethod('info.getTxFee')
                .then((response) => {
                return {
                    txFee: new bn_js_1.default(response.data.result.txFee, 10),
                    creationTxFee: new bn_js_1.default(response.data.result.creationTxFee, 10)
                };
            });
        });
        /**
         * Check whether a given chain is done bootstrapping
         * @param chain The ID or alias of a chain.
         *
         * @returns Returns a Promise<boolean> of whether the chain has completed bootstrapping.
         */
        this.isBootstrapped = (chain) => __awaiter(this, void 0, void 0, function* () {
            const params = {
                chain
            };
            return this.callMethod('info.isBootstrapped', params)
                .then((response) => response.data.result.isBootstrapped);
        });
        /**
         * Returns the peers connected to the node.
         *
         * @returns Promise for the list of connected peers in <ip>:<port> format.
         */
        this.peers = () => __awaiter(this, void 0, void 0, function* () {
            return this.callMethod('info.peers')
                .then((response) => response.data.result.peers);
        });
    }
}
exports.InfoAPI = InfoAPI;
