// avalanche/typings/tests/testlib.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestAPI = void 0;
const apibase_1 = require("../src/common/apibase");
class TestAPI extends apibase_1.APIBase {
    constructor(avax, endpoint = '/ext/testing') {
        super(avax, endpoint);
        this.TestGET = (input, path = '', axiosConfig = undefined) => __awaiter(this, void 0, void 0, function* () { return this._TestMethod('get', path, { input }, axiosConfig); });
        this.TestDELETE = (input, path = '', axiosConfig = undefined) => __awaiter(this, void 0, void 0, function* () { return this._TestMethod('delete', path, { input }, axiosConfig); });
        this.TestPOST = (input, path = '', axiosConfig = undefined) => __awaiter(this, void 0, void 0, function* () { return this._TestMethod('post', path, {}, { input }, axiosConfig); });
        this.TestPUT = (input, path = '', axiosConfig = undefined) => __awaiter(this, void 0, void 0, function* () { return this._TestMethod('put', path, {}, { input }, axiosConfig); });
        this.TestPATCH = (input, path = '', axiosConfig = undefined) => __awaiter(this, void 0, void 0, function* () { return this._TestMethod('patch', path, {}, { input }, axiosConfig); });
        this._respFn = (res) => {
            let response;
            if (typeof res.data === 'string') {
                response = JSON.parse(res.data);
            }
            else {
                response = res.data;
            }
            return response.result;
        };
        this._TestMethod = (method, path = '', getdata = {}, postdata = undefined, axiosConfig = undefined) => __awaiter(this, void 0, void 0, function* () {
            if (postdata === undefined) {
                return this.core[method](this.baseurl + path, getdata, {}, axiosConfig).then((res) => this._respFn(res));
            }
            return this.core[method](this.baseurl + path, getdata, postdata, {}, axiosConfig).then((res) => {
                res.data = JSON.stringify(res.data); // coverage completeness
                return this._respFn(res);
            });
        });
    }
}
exports.TestAPI = TestAPI;
