// avalanche/dist/common/apibase.js
"use strict";
/**
 * @packageDocumentation
 * @module Common-APIBase
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIBase = exports.RequestResponseData = void 0;
const bintools_1 = __importDefault(require("../utils/bintools"));
const db_1 = __importDefault(require("../utils/db"));
/**
 * @ignore
 */
const bintools = bintools_1.default.getInstance();
/**
 * Response data for HTTP requests.
 */
class RequestResponseData {
}
exports.RequestResponseData = RequestResponseData;
/**
 * Abstract class defining a generic endpoint that all endpoints must implement (extend).
 */
class APIBase {
    /**
       *
       * @param core Reference to the Avalanche instance using this baseurl
       * @param baseurl Path to the baseurl - ex: "/ext/bc/X"
       */
    constructor(core, baseurl) {
        /**
           * Sets the path of the APIs baseurl.
           *
           * @param baseurl Path of the APIs baseurl - ex: "/ext/bc/X"
           */
        this.setBaseURL = (baseurl) => {
            if (this.db && this.baseurl !== baseurl) {
                const backup = this.db.getAll();
                this.db.clearAll();
                this.baseurl = baseurl;
                this.db = db_1.default.getNamespace(baseurl);
                this.db.setAll(backup, true);
            }
            else {
                this.baseurl = baseurl;
                this.db = db_1.default.getNamespace(baseurl);
            }
        };
        /**
           * Returns the baseurl's path.
           */
        this.getBaseURL = () => this.baseurl;
        /**
           * Returns the baseurl's database.
           */
        this.getDB = () => this.db;
        this.core = core;
        this.setBaseURL(baseurl);
    }
}
exports.APIBase = APIBase;
