// avalanche/dist/apis/keystore/api.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeystoreAPI = void 0;
const jrpcapi_1 = require("../../common/jrpcapi");
/**
 * Class for interacting with a node API that is using the node's KeystoreAPI.
 *
 * **WARNING**: The KeystoreAPI is to be used by the node-owner as the data is stored locally on the node. Do not trust the root user. If you are not the node-owner, do not use this as your wallet.
 *
 * @category RPCAPIs
 *
 * @remarks This extends the [[JRPCAPI]] class. This class should not be directly called. Instead, use the [[Avalanche.addAPI]] function to register this interface with Avalanche.
 */
class KeystoreAPI extends jrpcapi_1.JRPCAPI {
    /**
       * This class should not be instantiated directly. Instead use the [[Avalanche.addAPI]] method.
       *
       * @param core A reference to the Avalanche class
       * @param baseurl Defaults to the string "/ext/keystore" as the path to blockchain's baseurl
       */
    constructor(core, baseurl = '/ext/keystore') {
        super(core, baseurl);
        /**
           * Creates a user in the node's database.
           *
           * @param username Name of the user to create
           * @param password Password for the user
           *
           * @returns Promise for a boolean with true on success
           */
        this.createUser = (username, password) => __awaiter(this, void 0, void 0, function* () {
            const params = {
                username,
                password,
            };
            return this.callMethod('keystore.createUser', params)
                .then((response) => response.data.result.success);
        });
        /**
           * Exports a user. The user can be imported to another node with keystore.importUser .
           *
           * @param username The name of the user to export
           * @param password The password of the user to export
           *
           * @returns Promise with a string importable using importUser
           */
        this.exportUser = (username, password) => __awaiter(this, void 0, void 0, function* () {
            const params = {
                username,
                password,
            };
            return this.callMethod('keystore.exportUser', params)
                .then((response) => response.data.result.user);
        });
        /**
           * Imports a user file into the node's user database and assigns it to a username.
           *
           * @param username The name the user file should be imported into
           * @param user cb58 serialized string represetning a user's data
           * @param password The user's password
           *
           * @returns A promise with a true-value on success.
           */
        this.importUser = (username, user, password) => __awaiter(this, void 0, void 0, function* () {
            const params = {
                username,
                user,
                password,
            };
            return this.callMethod('keystore.importUser', params)
                .then((response) => response.data.result.success);
        });
        /**
           * Lists the names of all users on the node.
           *
           * @returns Promise of an array with all user names.
           */
        this.listUsers = () => __awaiter(this, void 0, void 0, function* () {
            return this.callMethod('keystore.listUsers')
                .then((response) => response.data.result.users);
        });
        /**
           * Deletes a user in the node's database.
           *
           * @param username Name of the user to delete
           * @param password Password for the user
           *
           * @returns Promise for a boolean with true on success
           */
        this.deleteUser = (username, password) => __awaiter(this, void 0, void 0, function* () {
            const params = {
                username,
                password,
            };
            return this.callMethod('keystore.deleteUser', params)
                .then((response) => response.data.result.success);
        });
    }
}
exports.KeystoreAPI = KeystoreAPI;
