// avalanche/dist/apis/evm/keychain.js
"use strict";
/**
 * @packageDocumentation
 * @module API-EVM-KeyChain
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeyChain = exports.KeyPair = void 0;
const bintools_1 = __importDefault(require("../../utils/bintools"));
const secp256k1_1 = require("../../common/secp256k1");
/**
 * @ignore
 */
const bintools = bintools_1.default.getInstance();
/**
 * Class for representing a private and public keypair on an AVM Chain.
 */
class KeyPair extends secp256k1_1.SECP256k1KeyPair {
    constructor(hrp, chainid) {
        super();
        this.chainID = '';
        this.hrp = '';
        /**
         * Returns the address's string representation.
          *
          * @returns A string representation of the address
          */
        this.getAddressString = () => {
            const addr = this.addressFromPublicKey(this.pubk);
            return bintools.addressToString(this.hrp, this.chainID, addr);
        };
        /**
          * Returns the chainID associated with this key.
          *
          * @returns The [[KeyPair]]'s chainID
          */
        this.getChainID = () => this.chainID;
        /**
          * Sets the the chainID associated with this key.
          *
          * @param chainID String for the chainID
          */
        this.setChainID = (chainID) => {
            this.chainID = chainID;
        };
        /**
          * Returns the Human-Readable-Part of the network associated with this key.
          *
          * @returns The [[KeyPair]]'s Human-Readable-Part of the network's Bech32 addressing scheme
          */
        this.getHRP = () => this.hrp;
        /**
          * Sets the the Human-Readable-Part of the network associated with this key.
          *
          * @param hrp String for the Human-Readable-Part of Bech32 addresses
          */
        this.setHRP = (hrp) => {
            this.hrp = hrp;
        };
        this.chainID = chainid;
        this.hrp = hrp;
        this.generateKey();
    }
    clone() {
        let newkp = new KeyPair(this.hrp, this.chainID);
        newkp.importKey(bintools.copyFrom(this.getPrivateKey()));
        return newkp;
    }
    create(...args) {
        if (args.length == 2) {
            return new KeyPair(args[0], args[1]);
        }
        return new KeyPair(this.hrp, this.chainID);
    }
}
exports.KeyPair = KeyPair;
/**
  * Class for representing a key chain in Avalanche.
  *
  * @typeparam KeyPair Class extending [[SECP256k1KeyChain]] which is used as the key in [[KeyChain]]
  */
class KeyChain extends secp256k1_1.SECP256k1KeyChain {
    /**
      * Returns instance of KeyChain.
      */
    constructor(hrp, chainID) {
        super();
        this.hrp = '';
        this.chainID = '';
        /**
          * Makes a new key pair, returns the address.
          *
          * @returns The new key pair
          */
        this.makeKey = () => {
            let keypair = new KeyPair(this.hrp, this.chainID);
            this.addKey(keypair);
            return keypair;
        };
        this.addKey = (newKey) => {
            newKey.setChainID(this.chainID);
            super.addKey(newKey);
        };
        /**
          * Given a private key, makes a new key pair, returns the address.
          *
          * @param privk A {@link https://github.com/feross/buffer|Buffer}
          * or cb58 serialized string representing the private key
          *
          * @returns The new key pair
          */
        this.importKey = (privk) => {
            let keypair = new KeyPair(this.hrp, this.chainID);
            let pk;
            if (typeof privk === 'string') {
                pk = bintools.cb58Decode(privk.split('-')[1]);
            }
            else {
                pk = bintools.copyFrom(privk);
            }
            keypair.importKey(pk);
            if (!(keypair.getAddress().toString("hex") in this.keys)) {
                this.addKey(keypair);
            }
            return keypair;
        };
        this.hrp = hrp;
        this.chainID = chainID;
    }
    create(...args) {
        if (args.length == 2) {
            return new KeyChain(args[0], args[1]);
        }
        return new KeyChain(this.hrp, this.chainID);
    }
    ;
    clone() {
        const newkc = new KeyChain(this.hrp, this.chainID);
        for (let k in this.keys) {
            newkc.addKey(this.keys[k].clone());
        }
        return newkc;
    }
    ;
    union(kc) {
        let newkc = kc.clone();
        for (let k in this.keys) {
            newkc.addKey(this.keys[k].clone());
        }
        return newkc;
    }
}
exports.KeyChain = KeyChain;
